packages/slice-machine/components/Forms/CreateSliceModal/CreateSliceModal.tsx

```diff
@@ -1,62 +1,124 @@
+import { getState } from "@src/apiClient";
+import { CUSTOM_TYPES_MESSAGES } from "@src/features/customTypes/customTypesMessages";
+import { managerClient } from "@src/managerClient";
 import { validateSliceModalValues } from "../formsValidator";
-
-const formId = "create-new-slice";
+import { InputBox } from "../components/InputBox";
 
 type CreateSliceModalProps = {
-  isOpen: boolean;
-  isCreatingSlice: boolean;
-  onSubmit: ({ sliceName, from }: { sliceName: string; from: string }) => void;
-  close: () => void;
-  libraries: readonly LibraryUI[];
+  onClose: () => void;
+  localLibraries: readonly LibraryUI[];
   remoteSlices: ReadonlyArray<SliceSM>;
-  actionMessage?: ReactNode;
+  currentCustomType?: CustomTypeSM;
+  tabId?: string;
 };
 
 type FormValues = { sliceName: string; from: string };
 
-export const CreateSliceModal: React.FunctionComponent<
-  CreateSliceModalProps
-> = ({
-  isOpen,
-  isCreatingSlice,
-  onSubmit,
-  close,
-  libraries,
+export const CreateSliceModal: FC<CreateSliceModalProps> = ({
+  onClose,
+  localLibraries,
   remoteSlices,
-  actionMessage,
+  currentCustomType,
+  tabId,
 }) => {
+  const router = useRouter();
+  const { createSliceSuccess, saveCustomTypeSuccess } =
+    useSliceMachineActions();
+  const [isCreatingSlice, setIsCreatingSlice] = useState(false);
+
+  const onSubmit = async (values: FormValues) => {
+    const sliceName = values.sliceName;
+    const libraryName = values.from;
+
+    setIsCreatingSlice(true);
+
+    await createSlice({
+      sliceName,
+      libraryName,
+      onSuccess: async (newSlice) => {
+        // TODO(DT-1453): Remove the need of the global getState
+        const serverState = await getState();
+        // Update Redux store
+        createSliceSuccess(serverState.libraries);
+
+        // When creating a slice from a custom type, we add it directly to the slice zone and save
+        if (currentCustomType) {
+          const newCustomType = {
+            ...currentCustomType,
+            tabs: currentCustomType.tabs.map((tab) =>
+              tab.key === tabId && tab.sliceZone
+                ? {
+                    ...tab,
+                    sliceZone: {
+                      key: tab.sliceZone.key,
+                      value: [
+                        ...tab.sliceZone.value,
+                        {
+                          key: newSlice.id,
+                          value: newSlice,
+                        },
+                      ],
+                    },
+                  }
+                : tab
+            ),
+          };
+          await managerClient.customTypes.updateCustomType({
+            model: CustomTypes.fromSM(newCustomType),
+          });
+          saveCustomTypeSuccess(CustomTypes.fromSM(newCustomType));
+        }
+
+        // Redirect to the slice page
+        const variationId = newSlice.variations[0].id;
+        const sliceLocation = `/${libraryName.replace(
```

Comment:
I would prefer this to be an external function, a bit like you did for the CT tables